mac/Keyman4MacIM/Keyman4MacIM/KMInputMethodAppDelegate.m

```diff
@@ -1165,42 +1178,111 @@ - (void)postKeyboardEventWithSource: (CGEventSourceRef)source code:(CGKeyCode) v
     }
 }
 
+
+// TODO: This seriously needs to be refactored out of the app delegate and into
+//       a keyboard install module
 - (BOOL)unzipFile:(NSString *)filePath {
     BOOL didUnzip = NO;
+    NSError *error = nil;
     NSString *fileName = filePath.lastPathComponent;
     NSString *folderName = [fileName stringByDeletingPathExtension];
+
+    // First we unzip into a temp folder, and check kmp.json for the fileVersion
+    // before we continue installation. We don't want to overwrite existing
+    // package if it is there if the files are not compatible with the installed
+    // version of Keyman.
+
+    NSString *tempFolderName = [folderName stringByAppendingString:@".tmp.install"];
+    NSString *tempDestFolder = [self.keyboardsPath stringByAppendingPathComponent:tempFolderName];
+
     ZipArchive *za = [[ZipArchive alloc] init];
     if ([za UnzipOpenFile:filePath]) {
-        NSString *destFolder = [self.keyboardsPath stringByAppendingPathComponent:folderName];
         if (self.debugMode) {
-            NSLog(@"Unzipping %@ to %@", filePath, destFolder);
-            if ([[NSFileManager defaultManager] fileExistsAtPath:destFolder])
-                NSLog(@"The destiination folder already exists. Overwriting...");
+            NSLog(@"Unzipping %@ to %@", filePath, tempDestFolder);
+            if ([[NSFileManager defaultManager] fileExistsAtPath:tempDestFolder])
+                NSLog(@"The temp destination folder already exists. Overwriting...");
         }
-        didUnzip = [za UnzipFileTo:destFolder overWrite:YES];
+
+        didUnzip = [za UnzipFileTo:tempDestFolder overWrite:YES];
         [za UnzipCloseFile];
     }
 
-    if (didUnzip) {
-        if (self.debugMode)
-            NSLog(@"Unzipped file: %@", filePath);
-        NSString * keyboardFolderPath = [self.keyboardsPath stringByAppendingPathComponent:folderName];
-        [self installFontsAtPath:keyboardFolderPath];
-        for (NSString *kmxFile in [self KMXFilesAtPath:keyboardFolderPath]) {
-            if (self.debugMode)
-                NSLog(@"Adding keyboard to list of active keyboards: %@", kmxFile);
-            if (![self.activeKeyboards containsObject:kmxFile])
-                [self.activeKeyboards addObject:kmxFile];
+    if (!didUnzip) {
+        NSLog(@"Failed to unzip file: %@", filePath);
+        return NO;
+    }
+
+    if (self.debugMode)
+        NSLog(@"Unzipped file: %@", filePath);
+
+    BOOL didInstall = NO;
+
+    // Check the package info to ensure that we support this version
+    KMPackageInfo *packageInfo = [self loadPackageInfo:tempDestFolder];
+    if(packageInfo == nil) {
+        NSLog(@"Could not find kmp.json in %@", filePath);
+    } else {
+        NSString* requiredVersion = [packageInfo.fileVersion shortenedVersionNumberString];
```

Comment:
Is `packageInfo` the entire kmp.json file?
Would we be checking `packageInfo.system.fileVersion`?

Reply:
`packageInfo` is a translation of the file, not the raw json structure.